File: platform_src/ios/Elastos/Classes/PermissionManager.swift
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Foundation


class PluginPermission {
    var apiList = [String: Bool]();
    var defaultValue = true;

    init(_ defaultValue: Bool) {
        self.defaultValue = defaultValue;
    }

    func setDefaultValue(_ defaultValue: Bool) {
        self.defaultValue = defaultValue;
    }

    func getApiPermission(_ api: String) -> Bool {
        if (apiList.isEmpty) {
            return defaultValue;
        }

        let permission = apiList[api];
        if (permission == nil) {
            return defaultValue;
        }

        return permission!;
    }

    func setApiPermission(_ api: String, _ permission: Bool) {
        apiList[api] = permission;
    }
}

class PermissionGroup {

    var pluginList = [String: PluginPermission]();
    let name: String;

    init(_ name: String) {
        self.name = name;
    }

    func addPlugin(_ plugin: String, _ defaultValue: Bool) -> PluginPermission {
        var pluginPermission = pluginList[plugin];
        if (pluginPermission == nil) {
            pluginPermission = PluginPermission(defaultValue);
            pluginList[plugin] = pluginPermission;
        }
        return pluginPermission!;
    }

    func getApiPermission(_ plugin: String, _ api: String) -> Bool {
        let pluginPermission = pluginList[plugin];
        if (pluginPermission == nil) {
            return true;
        }

        return pluginPermission!.getApiPermission(api);
    }
}

class PermissionManager {
    var groupList = [String: PermissionGroup]();
    var appList = [String: String]();
    static var permissionManager: PermissionManager?;

    public init() {
        do {
            try parsePermissionGroup();
            try parseAppPermission();
        }
        catch let error {
            print("PermissionManager para config error: \(error)");
        }
        PermissionManager.permissionManager = self;
    }

    static func getShareInstance() -> PermissionManager {
        return PermissionManager.permissionManager!;
    }

    func parsePermissionGroup() throws {
        let path = getAbsolutePath("www/config/permission/group.json");
        let url = URL.init(fileURLWithPath: path)

        let data = try Data(contentsOf: url);
        let json = try JSONSerialization.jsonObject(with: data,
            options: []) as! [String: [String: [String]]];

        for (group, plugins) in json {
            var permissionGroup = PermissionGroup(group);
            groupList[group] = permissionGroup;
            for (var plugin, array) in plugins {
                var defaultValue = true;
                if (plugin.hasPrefix("+")) {
                    plugin = (plugin as NSString).substring(from: 1);
                    defaultValue = false;
                } else if (plugin.hasPrefix("-")) {
                    plugin = (plugin as NSString).substring(from: 1);
                }
                plugin = plugin.lowercased();

                var pluginPermission = permissionGroup.addPlugin(plugin, defaultValue);
                for i in 0..<array.count {
                    let api = array[i];
                    if (api == "*") {
                        pluginPermission.setDefaultValue(!defaultValue);
                        break;
                    }
                    pluginPermission.setApiPermission(api, !defaultValue);
                }
            }
        }
    }

    func parseAppPermission() throws {
        let path = getAbsolutePath("www/config/permission/app.json");
        let url = URL.init(fileURLWithPath: path)

        let data = try Data(contentsOf: url);
        let json = try JSONSerialization.jsonObject(with: data,
            options: []) as! [String: String];

        for (app, group) in json {
            appList[app] = group;
        }

    }

    func getPermissionGroup(_ app: String) -> PermissionGroup {
        var group = appList[app];
        if (group == nil) {
            group = "user";
        }

        var permissionGroup = groupList[group!];
        if (permissionGroup == nil) {
            permissionGroup = groupList["user"];
        }

        return permissionGroup!;
    }
}

File: platform_src/android/app/src/main/java/org/elastos/trinity/runtime/PermissionManager.java
package org.elastos.trinity.runtime;

import android.content.Context;
import android.content.res.AssetManager;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.InputStream;
import java.util.Iterator;
import java.util.LinkedHashMap;

public class PermissionManager {

    private Context context = null;
    protected LinkedHashMap<String, PermissionGroup> groupList = new LinkedHashMap();
    protected LinkedHashMap<String, String> appList = new LinkedHashMap();
    private static PermissionManager permissionManager;

    public PermissionManager(Context context) {
        this.context = context;
        try {
            parsePermissionGroup();
            parseAppPermission();
        } catch (Exception e) {
            e.printStackTrace();
        }
        PermissionManager.permissionManager = this;
    }

    public static PermissionManager getShareInstance() {
        return PermissionManager.permissionManager;
    }

    public LinkedHashMap<String, PermissionGroup> getGroupList() {
        return groupList;
    }

    public void parsePermissionGroup()  throws Exception {
        AssetManager manager = context.getAssets();
        InputStream inputStream = manager.open("www/config/permission/group.json");

        JSONObject json = Utility.getJsonFromFile(inputStream);

        Iterator groups = json.keys();
        while (groups.hasNext()) {
            String group = (String) groups.next();
            JSONObject jgroup = json.getJSONObject(group);
            Iterator plugins = jgroup.keys();
            PermissionGroup permissionGroup = new PermissionGroup(group);
            groupList.put(group, permissionGroup);
            while (plugins.hasNext()) {
                String plugin = (String) plugins.next();

                if (plugin.equals("*")) {
                    permissionGroup.setDefaultValue(true);
                    break;
                }

                if (plugin.startsWith("$")) {
                    permissionGroup.addBaseGroup(plugin.substring(1));
                    continue;
                }

                JSONArray array = jgroup.getJSONArray(plugin);
                Boolean defaultValue = false;
                if (plugin.startsWith("-")) {
                    plugin = plugin.substring(1);
                    defaultValue = true;
                }
                else if (plugin.startsWith("+")) {
                    plugin = plugin.substring(1);
                }

                PermissionGroup.PluginPermission pluginPermission = permissionGroup.addPlugin(plugin, defaultValue);

                for (int i = 0; i < array.length(); i++) {
                    String api = array.getString(i);
                    if (api.equals("*")) {
                        pluginPermission.setDefaultValue(!defaultValue);
                        break;
                    }
                    pluginPermission.setApiPermission(api, !defaultValue);
                }
            }
        }
    }

    public void parseAppPermission()  throws Exception {
        AssetManager manager = context.getAssets();
        InputStream inputStream = manager.open("www/config/permission/app.json");

        JSONObject json = Utility.getJsonFromFile(inputStream);

        Iterator apps = json.keys();
        while (apps.hasNext()) {
            String app = (String) apps.next();
            String group = json.getString(app);
            appList.put(app, group);
        }
    }

    public PermissionGroup getPermissionGroup(String app) {
        String group = appList.get(app);
        if (group == null) {
            group = "user";
        }

        PermissionGroup permissionGroup = groupList.get(group);
        if (permissionGroup == null) {
            permissionGroup = groupList.get("user");
        }

        return permissionGroup;
    }
}
